R/serializer.R

```diff
@@ -558,68 +573,142 @@ serializer_device <- function(type, dev_on, dev_off = grDevices::dev.off) {
       } else {
         success(result)
       }
-    }
+    },
+    serializer_params = serializer_params
   )
 }
 
+serializer_param_width <- function() {
+  list(plot_width = list(desc="Width of plot image in units", type="number", required=FALSE, isArray=FALSE))
+}
+serializer_param_height <- function() {
+  list(plot_height = list(desc="Height of plot image in units", type="number", required=FALSE, isArray=FALSE))
+}
+serializer_param_res <- function() {
+  list(plot_res = list(desc="Resolution of plot image", type="number", required=FALSE, isArray=FALSE))
+}
+serializer_param_units <- function() {
+  list(plot_units = list(desc="Units of plot image", type="string", required=FALSE, isArray=FALSE))
+}
+serializer_param_pointsize <- function() {
+  list(plot_pointsize = list(desc="Point size of plot image - TODO - better desc", type="number", required=FALSE, isArray=FALSE))
+}
+serializer_param_bg <- function() {
+  list(plot_bg = list(desc="Background colour of plot image", type="string", required=FALSE, isArray=FALSE))
+}
+
+serializer_param_list <- function(all_except = character()) {
+  params <- c(
+    serializer_param_width(),
+    serializer_param_height(),
+    serializer_param_res(),
+    serializer_param_units(),
+    serializer_param_pointsize(),
+    serializer_param_bg()
+  )
+
+  # oh I wish I had purrr!
+  ignored <- Map(function(x) {
+    params[[x]] <<- NULL
+  }, all_except)
+
+  params
+}
+
+
+serialize_dimensions_args_preparer <- function(req, ...) {
+
+  # if legacy plumber.staticSerializers is specified then ignore all req
+  # based parameters
+  if (getOption("staticSerializers", default="FALSE") == "TRUE") {
+    return(list(...))
+  }
+
+  as_numeric_nullable <- function(preferred, alternative) {
+    option <- preferred %||% alternative
+    if (is.null(option)) {
+      NULL
+    } else {
+      as.numeric(option)
+    }
+  }
+
+  doc_args <- list(...)
+  doc_args$width <- as_numeric_nullable(req$args$plot_width, doc_args$width)
+  doc_args$height <- as_numeric_nullable(req$args$plot_height, doc_args$height)
+  doc_args$units <- req$args$plot_units %||% doc_args$units
+  doc_args$res <- as_numeric_nullable(req$args$plot_res, doc_args$res)
+  doc_args$pointsize <- as_numeric_nullable(req$args$plot_pointsize, doc_args$pointsize)
+  doc_args$bg <- req$args$plot_bg %||% doc_args$bg
```

Comment:
What is your motivation to use `plot_*` for the arg names? 
Why not pass them through directly?

My interpretation:
* Minimal param clashing with `plot_*`
* Confusion of two differently named parameters
	* Params in `@serializer png list(width=200)`
	* Params in `http://....../plot_route?plot_width=200`

If there isn't too much motivation for using `plot_*` names, I'd like to keep the consistent arg names when possible. (Knowing there could be param name clashing with the regular route params)

Reply:
As well as avoiding clashes, my main motivation was `plot_` names was to make it easier to distinguish the plot parameters from the business logic - e.g.  in https://.../palmerPenguins?min_flipper_length=200&min_bill_length=40&plot_width=400&plot_height=500 


The naming also makes the business logic vs plot parameter separation obvious in the Swagger UI (maybe some more ordering there might also help?)

e.g. it's obvious here which params are business vs which are plot:
![image](https://user-images.githubusercontent.com/533662/210643852-60e90e6f-df5a-4dd1-867c-9f351ce2d84e.png)

Asides: 
- initially I was wondering about using e.g. `plot.width` to try to make the plot parameters look like a separate object in the REST url (but couldn't spot any docs on that in https://swagger.io/docs/specification/describing-parameters/)
- also wondering whether we should allow HTTP header insertion of these parameters (not sure that's very RESTFUL though)